examples/cpp/gcp_observability/helloworld/greeter_server.cc

```diff
@@ -0,0 +1,95 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#include <iostream>
+#include <memory>
+#include <string>
+
+#include "absl/flags/flag.h"
+#include "absl/flags/parse.h"
+#include "absl/strings/str_format.h"
+
+#include <grpcpp/ext/gcp_observability.h>
+#include <grpcpp/ext/proto_server_reflection_plugin.h>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/health_check_service_interface.h>
+
+#ifdef BAZEL_BUILD
+#include "examples/protos/helloworld.grpc.pb.h"
+#else
+#include "helloworld.grpc.pb.h"
+#endif
+
+using grpc::Server;
+using grpc::ServerBuilder;
+using grpc::ServerContext;
+using grpc::Status;
+using helloworld::Greeter;
+using helloworld::HelloReply;
+using helloworld::HelloRequest;
+
+ABSL_FLAG(uint16_t, port, 50051, "Server port for the service");
+
+// Logic and data behind the server's behavior.
+class GreeterServiceImpl final : public Greeter::Service {
+  Status SayHello(ServerContext* context, const HelloRequest* request,
+                  HelloReply* reply) override {
+    std::string prefix("Hello ");
+    reply->set_message(prefix + request->name());
+    return Status::OK;
+  }
+};
+
+void RunServer(uint16_t port) {
+  std::string server_address = absl::StrFormat("0.0.0.0:%d", port);
+  GreeterServiceImpl service;
+  grpc::EnableDefaultHealthCheckService(true);
+  grpc::reflection::InitProtoReflectionServerBuilderPlugin();
+  ServerBuilder builder;
+  // Listen on the given address without any authentication mechanism.
+  builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
+  // Register "service" as the instance through which we'll communicate with
+  // clients. In this case it corresponds to an *synchronous* service.
+  builder.RegisterService(&service);
+  // Finally assemble the server.
+  std::unique_ptr<Server> server(builder.BuildAndStart());
+  std::cout << "Server listening on " << server_address << std::endl;
+
+  // Wait for the server to shutdown. Note that some other thread must be
+  // responsible for shutting down the server for this call to ever return.
+  server->Wait();
+}
+
+int main(int argc, char** argv) {
+  absl::ParseCommandLine(argc, argv);
+  // Turn on GCP Observability for the whole binary. Note that this should be
+  // done before any other gRPC operation.
+  auto status = grpc::experimental::GcpObservabilityInit();
+  if (!status.ok()) {
+    std::cerr << "GcpObservabilityInit() failed: " << status.ToString()
+              << std::endl;
+    return static_cast<int>(status.code());
+  }
+  std::cout << "Initialized GCP Observability" << std::endl;
+  RunServer(absl::GetFlag(FLAGS_port));
+  // Note that RunServer() never returns and hence there is no need for a
+  // GcpObservabilityClose() here, but as on the client side, we need to wait
+  // atleast 25seconds after the last RPC to make sure that stats and tracing
+  // are flushed.
```

Comment:
See the golang example. We catch ctrl c signals (and another exit signal), block on that (since that's only way to kill server binary) and then use that even to trigger server.Stop() exiting. server.Stop() exiting causes main() to finish and thus run our o11y cleanup (which flushes all 3) in a defer from main.

Reply:
Alright, I'll take your comment.. was feeling lazy earlier.